### PR TITLE
[SYCL] Implement sycl_ext_oneapi_address_cast

### PR DESCRIPTION
**Pull request by Pennycook**
- Adds support for OpGenericCastToPtr SPIR-V intrinsic.
- Implements static_address_cast using OpGenericCastToPtr.
- Implements dynamic_address_cast using OpGenericCastToPtrExplicit.
- Adds tests for both new forms of address_cast.

### CODE REVIEW

File: sycl/include/CL/__spirv/spirv_ops.hpp [hunk 1]
```diff
@@ -467,6 +467,94 @@ __SYCL_GenericCastToPtrExplicit_ToPrivate(const volatile void *Ptr) noexcept {
                                                  __spv::StorageClass::Function);
 }
 
+template <typename dataT>
+extern __attribute__((opencl_global)) dataT *
+__SYCL_GenericCastToPtr_ToGlobal(void *Ptr) noexcept {
+  return (__attribute__((opencl_global)) dataT *)
```
Comment (Naghasan) [inline sycl/include/CL/__spirv/spirv_ops.hpp:473]: minor: no need for cast, the returned type should be already correctly typed

**Review by Naghasan — APPROVED** [review]

File: sycl/test-e2e/AddressCast/dynamic_address_cast.cpp [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+                        sycl::access::address_space::local_space,
+                        sycl::access::decorated::no>(RawGlobalPointer);
+                auto PrivatePointer =
+                    sycl::ext::oneapi::experimental::dynamic_address_cast<
+                        sycl::access::address_space::private_space,
+                        sycl::access::decorated::no>(RawGlobalPointer);
+                Success &= reinterpret_cast<size_t>(RawGlobalPointer) ==
+                           reinterpret_cast<size_t>(GlobalPointer.get_raw());
+                Success &= LocalPointer.get_raw() == nullptr;
+                Success &= PrivatePointer.get_raw() == nullptr;
```
Comment (Naghasan) [inline sycl/test-e2e/AddressCast/dynamic_address_cast.cpp:54]: @mmoadeli for your awareness, your patch is going to be required to enable this test on cuda and hip

**Review by sergey-semenov — APPROVED** [review]

**Review by premanandrao — APPROVED** [review]
FE changes LGTM.

**Comment by Pennycook** [comment]
@intel/llvm-gatekeepers, I think this can be merged now.

**Pull request merged**